src/dstore/dstore_base.c

```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
 
 	return rc;
 }
+
+int dstore_dealloc_op(struct dstore_obj *obj, struct dstore_io_vec *vec,
+		      struct dstore_io_op **out)
+{
+	int rc;
+
+	rc = dstore_io_op_init_and_submit(obj, vec, out,
+					  DSTORE_IO_OP_FREE);
+
+	log_debug("dstore_dealloc_op (" OBJ_ID_F " <=> %p, "
+		  "vec=%p *out=%p) rc=%d",
+		  OBJ_ID_P(dstore_obj_id(obj)), obj, vec,
+		  rc == 0 ? *out : NULL, rc);
+
+	return rc;
+}
+
+int dstore_deallocate(struct dstore_obj *obj, off_t offset, size_t count)
+{
+	int rc = 0;
+
+	dassert(obj);
+
+	char *tmp_buf = NULL;
+	struct dstore_io_op *wop = NULL;
+	struct dstore_io_vec *vec = NULL;
+	size_t nr_request;
+	size_t tail_size;
+	size_t nblk_per_req;
+	size_t ndata_per_req;
+	int i;
+
+	size_t bsize = dstore_get_bsize(obj->ds,
+					(dstore_oid_t *)dstore_obj_id(obj));
+
+	if (offset % bsize != 0) {
+		/*we need to make deallocate operation left aligned */
+		tmp_buf = calloc(bsize, sizeof(char));
+		uint32_t left_blk_num = offset/bsize;
+		uint32_t write_count = offset - (left_blk_num * bsize);
+		write_count = bsize - write_count;
+
+		if (count < write_count) {
+
+			write_count = count;
+		}
+
+		RC_WRAP_LABEL(rc, out, dstore_pwrite, obj,
+			      offset, write_count,
+			      bsize, tmp_buf);
+
+		if (write_count == count) {
+			goto out;
+		}
+
+		count = count - write_count;
+		offset = offset + write_count;
+	}
+
+	ndata_per_req = DSAL_MAX_IO_SIZE;	
+	nr_request = count / ndata_per_req;
+	tail_size = count - (nr_request * ndata_per_req);
+	nblk_per_req = ndata_per_req / bsize;
+
+	vec = calloc(1, sizeof(struct dstore_io_vec));
+        if (vec == NULL) {
+                rc = -ENOMEM;
+                goto out;
+        }
+	/* we do not need io buffer here, we just need to send extents info */
+	vec->flags |= IO_NO_DATA;
+	dstore_io_vec_set_from_edbuf(vec);
+
+	for (i = 0; i < nr_request; i++) {
+
+		*(vec->ovec) = offset;
+		*(vec->svec) = ndata_per_req;
+
+		RC_WRAP_LABEL(rc, out, dstore_dealloc_op, obj, vec, &wop);
+
+		RC_WRAP_LABEL(rc, out, dstore_io_op_wait, wop);
+
+		dstore_io_op_fini(wop);
+
+		wop = NULL;
+
+		offset += ndata_per_req;
+	}
+
+	if (tail_size) {
+
+		*(vec->ovec) = offset;
+		*(vec->svec) = tail_size;
+
+		RC_WRAP_LABEL(rc, out, dstore_dealloc_op, obj, vec, &wop);
+
+		RC_WRAP_LABEL(rc, out, dstore_io_op_wait, wop);
+
+		dstore_io_op_fini(wop);
+
+		wop = NULL;
+
+		offset += tail_size;
+	}
+
+out:
+	if (wop) {
+		dstore_io_op_fini(wop);
+	}
+
+	if (tmp_buf) {
+		free(tmp_buf);
+	}
+
+	log_trace("dstore_deallocate:(" OBJ_ID_F " <=> %p )"
+                  "offset = %lu count = %lu rc = %d",
+                  OBJ_ID_P(dstore_obj_id(obj)), obj, offset, count, rc);
```

Comment:
If you see line number 862 and 846, you are modifying `offset` and then logging it. So ultimately you are not logging the original input `offset` but modified one and it may mislead. Let's say you came here with `rc = 0` it means all operation completed successfully then at this place your `count` and `offset` will be same and we will be wondering how did we reached here if we have `count == offset`.

Reply:
Done.